In [1]:
import numpy as np
import pandas as pd

In [3]:
data = pd.read_csv('C:\\Users\\Dell\\Python practise\\one hot encoding\\test.csv',usecols=['X1','X2','X3','X4','X5','X6'])
data.head()

X1  X2 X3 X4 X5 X6
0  v   n  f  d  t  a
1  b  ai  a  d  b  g
2  v  as  f  d  a  j
3  l   n  f  d  z  l
4  s  as  c  d  y  i

In [4]:
#to check how many labels are there
for col in data.columns:
    print(col,':',len(data[col].unique()),'labels')

X1 : 27 labels
X2 : 45 labels
X3 : 7 labels
X4 : 4 labels
X5 : 32 labels
X6 : 12 labels


In [8]:
#now if we apply one hot encoding how many columns will be created
pd.get_dummies(data,drop_first=True).shape

(4209, 121)

we can see just for 6 columns the encoding leads to 121 columns thus this leads to curse of dimenstionality i.e leads to error in accuracy. to fix this see below

In [10]:
#performing one hot encoding for top 10 most frequent category
data.X1.value_counts().sort_values(ascending=False).head(20)

aa    826
s     602
l     599
b     596
v     436
r     252
i     189
a     153
c     142
o      81
w      50
u      40
z      31
e      29
m      27
h      27
j      22
y      21
t      18
n      16
Name: X1, dtype: int64

In [11]:
#similarly for other columns like X1 X3 X4... can be performed taking for X2
#performing one hot encoding for top 10 most frequent category
data.X2.value_counts().sort_values(ascending=False).head(20)

as    1658
ae     478
ai     462
m      348
ak     260
r      155
n      113
s      100
f       85
e       84
ay      78
aq      72
a       44
b       38
t       25
k       25
ag      23
ac      20
ao      19
i       15
Name: X2, dtype: int64

In [16]:
top_10 =[x for x in data.X2.value_counts().sort_values(ascending=False).head(10).index]
top_10

['as', 'ae', 'ai', 'm', 'ak', 'r', 'n', 's', 'f', 'e']

In [18]:
for label in top_10:
    data[label]=np.where(data['X2']==label,1,0)

data[['X2']+top_10].head(40)    

X2  as  ae  ai  m  ak  r  n  s  f  e
0    n   0   0   0  0   0  0  1  0  0  0
1   ai   0   0   1  0   0  0  0  0  0  0
2   as   1   0   0  0   0  0  0  0  0  0
3    n   0   0   0  0   0  0  1  0  0  0
4   as   1   0   0  0   0  0  0  0  0  0
5   ai   0   0   1  0   0  0  0  0  0  0
6   ae   0   1   0  0   0  0  0  0  0  0
7   ae   0   1   0  0   0  0  0  0  0  0
8    s   0   0   0  0   0  0  0  1  0  0
9   as   1   0   0  0   0  0  0  0  0  0
10   s   0   0   0  0   0  0  0  1  0  0
11   b   0   0   0  0   0  0  0  0  0  0
12   e   0   0   0  0   0  0  0  0  0  1
13  ae   0   1   0  0   0  0  0  0  0  0
14  ak   0   0   0  0   1  0  0  0  0  0
15   e   0   0   0  0   0  0  0  0  0  1
16   s   0   0   0  0   0  0  0  1  0  0
17   e   0   0   0  0   0  0  0  0  0  1
18  ai   0   0   1  0   0  0  0  0  0  0
19  as   1   0   0  0   0  0  0  0  0  0
20   e   0   0   0  0   0  0  0  0  0  1
21   e   0   0   0  0   0  0  0  0  0  1
22  ai   0   0   1  0   0  0  0  0  0  0
23   e   0   0   0  0   0  0  0  0  0  1
24  ai   0   0   1  0   0  0  0  0  0  0
25   s   0   0   0  0   0  0  0  1  0  0
26   m   0   0   0  1   0  0  0  0  0  0
27  as   1   0   0  0   0  0  0  0  0  0
28   m   0   0   0  1   0  0  0  0  0  0
29  as   1   0   0  0   0  0  0  0  0  0
30  as   1   0   0  0   0  0  0  0  0  0
31  as   1   0   0  0   0  0  0  0  0  0
32  as   1   0   0  0   0  0  0  0  0  0
33   a   0   0   0  0   0  0  0  0  0  0
34  aq   0   0   0  0   0  0  0  0  0  0
35  as   1   0   0  0   0  0  0  0  0  0
36  ae   0   1   0  0   0  0  0  0  0  0
37  ak   0   0   0  0   1  0  0  0  0  0
38  ag   0   0   0  0   0  0  0  0  0  0
39  ae   0   1   0  0   0  0  0  0  0  0

In [21]:
#for all categorical variables
def one_hot_top(df,variable,top_x_labels):
    for label in top_x_labels:
        df[variable +'_'+label]=np.where(data[variable]==label,1,0)
#read data again        
data = pd.read_csv('C:\\Users\\Dell\\Python practise\\one hot encoding\\test.csv',usecols=['X1','X2','X3','X4','X5','X6'])

#top 10 frequent for x1 
top_10=[x for x in data.X1.value_counts().sort_values(ascending=False).head(10).index]

#encode X1 into 10 most frequent categories
one_hot_top(data,'X1',top_10)
data.head()

X1  X2 X3 X4 X5 X6  X1_aa  X1_s  X1_l  X1_b  X1_v  X1_r  X1_i  X1_a  X1_c  \
0  v   n  f  d  t  a      0     0     0     0     1     0     0     0     0   
1  b  ai  a  d  b  g      0     0     0     1     0     0     0     0     0   
2  v  as  f  d  a  j      0     0     0     0     1     0     0     0     0   
3  l   n  f  d  z  l      0     0     1     0     0     0     0     0     0   
4  s  as  c  d  y  i      0     1     0     0     0     0     0     0     0   

   X1_o  
0     0  
1     0  
2     0  
3     0  
4     0

In [22]:
#top 10 frequent for x2
top_10=[x for x in data.X2.value_counts().sort_values(ascending=False).head(10).index]

#encode X2 into 10 most frequent categories
one_hot_top(data,'X2',top_10)
data.head()

X1  X2 X3 X4 X5 X6  X1_aa  X1_s  X1_l  X1_b  ...  X2_as  X2_ae  X2_ai  X2_m  \
0  v   n  f  d  t  a      0     0     0     0  ...      0      0      0     0   
1  b  ai  a  d  b  g      0     0     0     1  ...      0      0      1     0   
2  v  as  f  d  a  j      0     0     0     0  ...      1      0      0     0   
3  l   n  f  d  z  l      0     0     1     0  ...      0      0      0     0   
4  s  as  c  d  y  i      0     1     0     0  ...      1      0      0     0   

   X2_ak  X2_r  X2_n  X2_s  X2_f  X2_e  
0      0     0     1     0     0     0  
1      0     0     0     0     0     0  
2      0     0     0     0     0     0  
3      0     0     1     0     0     0  
4      0     0     0     0     0     0  

[5 rows x 26 columns]